### PR TITLE
Update reactnativeappbuilder

### PR DESCRIPTION
**Pull request by sharath2727**
### Type of Change
- Fabric Application Changes

### Why
Make the Win32 Fabric App less cumbersome to get started with.

Resolves [Add Relevant Issue Here]

### What
Added a new runtimeclass ReactNativeWin32App and made changes to ReactNativeAppBuilder functionality

## Changelog
no_

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/react-native-windows/pull/13467)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp [hunk 1]
```diff
@@ -38,29 +43,69 @@ winrt::ReactNative::ReactNativeAppBuilder ReactNativeAppBuilder::SetReactInstanc
...
+  m_reactViewOptions = reactViewOptions;
 
   return *this;
 }
 
-winrt::ReactNative::ReactNativeHost ReactNativeAppBuilder::Start() {
-  auto hwnd{winrt::UI::GetWindowFromWindowId(m_appWindow.Id())};
+winrt::ReactNative::ReactNativeWin32App ReactNativeAppBuilder::Build() {
+  if (m_reactNativeWin32App.Compositor() == nullptr) {
+    m_reactNativeWin32App.as<implementation::ReactNativeWin32App>().get()->Compositor(nullptr);
```
Comment (jonthysell) [inline vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp:66]: Not really a fan of the setting it to nullptr to tell it to create it's own. Just create it here.

File: vnext/Microsoft.ReactNative/ReactNativeWin32App.cpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  m_appWindow = appWindow;
+}
+
+winrt::Microsoft::UI::Windowing::AppWindow ReactNativeWin32App::AppWindow() {
+  return m_appWindow;
+}
+
+void ReactNativeWin32App::Compositor(winrt::Microsoft::UI::Composition::Compositor const &compositor) {
+  if (compositor == nullptr) {
+    m_compositor = winrt::Microsoft::UI::Composition::Compositor();
```
Comment (jonthysell) [inline vnext/Microsoft.ReactNative/ReactNativeWin32App.cpp:31]: I don't think this logic should be here. If they pass in nullptr, set it to nullptr. Let build set it to something new if necessary.

File: vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp [hunk 3]
```diff
@@ -17,8 +18,12 @@ using namespace winrt::Microsoft::UI;
 namespace winrt::Microsoft::ReactNative::implementation {
 ReactNativeAppBuilder::ReactNativeAppBuilder() {
   m_host = winrt::make<winrt::ReactNative::implementation::ReactNativeHost>();
```
Comment (jonthysell) [inline vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp:20]: Why is the host still created here? You're already storing everything else in the ReactNativeWin32App, just let it exist there.

File: vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp [hunk 4]
```diff
@@ -38,29 +43,69 @@ winrt::ReactNative::ReactNativeAppBuilder ReactNativeAppBuilder::SetReactInstanc
...
 
   winrt::ReactNative::ReactCoreInjection::SetTopLevelWindowId(
       m_host.InstanceSettings().Properties(), reinterpret_cast<uint64_t>(hwnd));
 
-  winrt::ReactNative::Composition::CompositionUIService::SetCompositor(m_host.InstanceSettings(), m_compositor);
+  winrt::ReactNative::Composition::CompositionUIService::SetCompositor(
+      m_host.InstanceSettings(), m_reactNativeWin32App.Compositor());
 
   // Start the react-native instance, which will create a JavaScript runtime and load the applications bundle.
   m_host.ReloadInstance();
```
Comment (jonthysell) [inline vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp:87]: Is it possible this should just be in Start and not Build?

File: packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp [hunk 5]
```diff
@@ -90,27 +89,25 @@ _Use_decl_annotations_ int CALLBACK WinMain(HINSTANCE instance, HINSTANCE, PSTR
...
   // Start the react-native instance by creating a javascript runtime and load the bundle.
-  auto host{reactNativeAppBuilder.Start()};
+  auto reactNativeWin32App{reactNativeAppBuilder.Build()};
 
-  // Create a RootView which will present a react-native component
-  winrt::Microsoft::ReactNative::ReactViewOptions viewOptions;
-  viewOptions.ComponentName(mainComponentName);
-  auto rootView = winrt::Microsoft::ReactNative::ReactNativeIsland(compositor);
-  rootView.ReactViewHost(winrt::Microsoft::ReactNative::ReactCoreInjection::MakeViewHost(host, viewOptions));
+  auto reactNativeIsland{reactNativeWin32App.ReactNativeIsland()};
```
Comment (jonthysell) [inline packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp:107]: Everything from this line to Start should be in either Build or Start.

File: vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp [hunk 6]
```diff
@@ -38,29 +43,69 @@ winrt::ReactNative::ReactNativeAppBuilder ReactNativeAppBuilder::SetReactInstanc
...
-winrt::ReactNative::ReactNativeHost ReactNativeAppBuilder::Start() {
-  auto hwnd{winrt::UI::GetWindowFromWindowId(m_appWindow.Id())};
+winrt::ReactNative::ReactNativeWin32App ReactNativeAppBuilder::Build() {
+  if (m_reactNativeWin32App.Compositor() == nullptr) {
+    m_reactNativeWin32App.as<implementation::ReactNativeWin32App>().get()->Compositor(nullptr);
+  }
+
+  if (m_reactNativeWin32App.AppWindow() == nullptr) {
+    auto appWindow = winrt::Microsoft::UI::Windowing::AppWindow::Create();
+    appWindow.Title(L"Sample-App-Fabric");
```
Comment (jonthysell) [inline vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp:71]: Title shouldn't be set here (arguably also the starting size). This is a case where the title (and starting size) should be set by the app template in-between Build() and Start().

File: vnext/Microsoft.ReactNative/ReactNativeAppBuilder.h [hunk 7]
```diff
@@ -14,12 +17,14 @@ struct ReactNativeAppBuilder : ReactNativeAppBuilderT<ReactNativeAppBuilder> {
...
-  winrt::Microsoft::ReactNative::ReactNativeHost Start();
+  winrt::Microsoft::ReactNative::ReactNativeAppBuilder SetReactViewOptions(
+      winrt::Microsoft::ReactNative::ReactViewOptions const &reactViewOptions);
+  winrt::Microsoft::ReactNative::ReactNativeWin32App Build();
 
  private:
-  winrt::Microsoft::UI::Windowing::AppWindow m_appWindow{nullptr};
-  winrt::Microsoft::UI::Composition::Compositor m_compositor{};
   winrt::Microsoft::ReactNative::ReactNativeHost m_host{};
+  winrt::Microsoft::ReactNative::ReactViewOptions m_reactViewOptions{};
```
Comment (jonthysell) [inline vnext/Microsoft.ReactNative/ReactNativeAppBuilder.h:26]: Why is this here and not in ReactNativeWin32App?

File: vnext/Microsoft.ReactNative/ReactNativeWin32App.cpp [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+winrt::Microsoft::ReactNative::ReactNativeHost ReactNativeWin32App::ReactNativeHost() {
+  return m_host;
+}
+
+void ReactNativeWin32App::ReactNativeHost(winrt::Microsoft::ReactNative::ReactNativeHost const &host) {
+  m_host = host;
+}
+
+void ReactNativeWin32App::Start() {
+  m_desktopChildSiteBridge.Show();
```
Comment (jonthysell) [inline vnext/Microsoft.ReactNative/ReactNativeWin32App.cpp:67]: All the show and runevent loop should be in here.

File: packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp (hunk 5 above)
Comment (sharath2727) [reply packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp:101]: As discussed, For version1 we are going to continue with this pattern,

File: vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp (hunk 6 above)
Comment (sharath2727) [reply vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp:86]: But it shouldn't matter I have checked that the title can be changed any time before Start is called and it still reflects the right title.

File: vnext/Microsoft.ReactNative/ReactNativeAppBuilder.h (hunk 7 above)
Comment (sharath2727) [reply vnext/Microsoft.ReactNative/ReactNativeAppBuilder.h:27]: For V1, Let this be here. I need to still figure out some of these dependencies on when I can set the reactViewOptions.

File: vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp (hunk 6 above)
Comment (jonthysell) [reply vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp:86]: Yes, but the app builder is for all apps, not just "Sample-App-Fabric". Either remove the line or put some better default title string

File: packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp [hunk 9]
```diff
@@ -90,27 +83,25 @@ _Use_decl_annotations_ int CALLBACK WinMain(HINSTANCE instance, HINSTANCE, PSTR
...
 
   // Update the size of the RootView when the AppWindow changes size
-  window.Changed([wkRootView = winrt::make_weak(rootView)](
+  window.Changed([wkRootView = winrt::make_weak(reactNativeIsland)](
```
Comment (jonthysell) [inline packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp:104]: The window.Changed stuff should all be happening in Build().

File: packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp [hunk 10]
```diff
@@ -120,6 +111,8 @@ _Use_decl_annotations_ int CALLBACK WinMain(HINSTANCE instance, HINSTANCE, PSTR
     }
   });
 
+  auto host{reactNativeWin32App.ReactNativeHost()};
+
   // Quit application when main window is closed
   window.Destroying(
```
Comment (jonthysell) [inline packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp:117]: This should also be in Build().

File: packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp [hunk 11]
```diff
@@ -132,29 +125,8 @@ _Use_decl_annotations_ int CALLBACK WinMain(HINSTANCE instance, HINSTANCE, PSTR
...
-  // Run the main application event loop
-  dispatcherQueueController.DispatcherQueue().RunEventLoop();
-
-  // Rundown the DispatcherQueue. This drains the queue and raises events to let components
-  // know the message loop has finished.
-  dispatcherQueueController.ShutdownQueue();
-
-  bridge.Close();
-  bridge = nullptr;
+  UpdateRootViewSizeToAppWindow(reactNativeIsland, window);
```
Comment (jonthysell) [inline packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp:129]: This should be in Build()

File: packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp [hunk 12]
```diff
@@ -20,6 +20,7 @@ struct CompReactPackageProvider
 constexpr PCWSTR windowTitle = L"sample_app_fabric";
 constexpr PCWSTR mainComponentName = L"sample_app_fabric";
 
+// Scaling factor for the window's content based on the DPI of the display where the window is located.
 float ScaleFactor(HWND hwnd) noexcept {
```
Comment (jonthysell) [inline packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp:24]: This is no longer necessary since it's handled in Build

File: packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp [hunk 13]
```diff
@@ -47,20 +48,12 @@ _Use_decl_annotations_ int CALLBACK WinMain(HINSTANCE instance, HINSTANCE, PSTR
...
 
-  // Create a DispatcherQueue for this thread.  This is needed for Composition, Content, and
-  // Input APIs.
-  auto dispatcherQueueController{winrt::Microsoft::UI::Dispatching::DispatcherQueueController::CreateOnCurrentThread()};
-
-  // Create a Compositor for all Content on this thread.
-  auto compositor{winrt::Microsoft::UI::Composition::Compositor()};
-
   // Create a top-level window.
   auto window = winrt::Microsoft::UI::Windowing::AppWindow::Create();
```
Comment (jonthysell) [inline packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp:52]: The app builder is already handling the window creation, right? By default most users won't be creating and setting app window themselves

**Review by jonthysell — CHANGES_REQUESTED** [review]

File: vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp (hunk 6 above)
Comment (sharath2727) [reply vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp:86]: Maybe this would be a better string "SampleFabricApplication"

File: packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp (hunk 13 above)
Comment (sharath2727) [reply packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp:52]: This is to showcase that they can create a window and pass it to the builder.

File: vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp [hunk 14]
```diff
@@ -16,51 +39,138 @@ using namespace winrt::Microsoft::UI;
...
+
+    // Create the compositor on behalf of the App Developer
+    auto compositor = winrt::Microsoft::UI::Composition::Compositor();
+    m_reactNativeWin32App.as<implementation::ReactNativeWin32App>().get()->Compositor(compositor);
+  }
+
+  // Create the AppWindow if the developer doesn't provide one
+  if (m_reactNativeWin32App.AppWindow() == nullptr) {
+    auto appWindow = winrt::Microsoft::UI::Windowing::AppWindow::Create();
+    appWindow.Title(L"SampleFabricApplication");
```
Comment (acoates-ms) [inline vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp:102]: Pretty sure most apps wont want this to be their title.

File: vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp [hunk 15]
```diff
@@ -16,51 +39,138 @@ using namespace winrt::Microsoft::UI;
...
+    appWindow.Resize({1000, 1000});
```
Comment (acoates-ms) [inline vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp:103]: Can apps specify a custom size?

File: vnext/Microsoft.ReactNative/ReactNativeAppBuilder.h [hunk 16]
```diff
@@ -1,25 +1,31 @@
...
 
+  ~ReactNativeAppBuilder();
+
   winrt::Microsoft::ReactNative::ReactNativeAppBuilder AddPackageProviders(
       winrt::Windows::Foundation::Collections::IVector<winrt::Microsoft::ReactNative::IReactPackageProvider> const
           &packageProviders);
   winrt::Microsoft::ReactNative::ReactNativeAppBuilder SetReactInstanceSettings(
       winrt::Microsoft::ReactNative::ReactInstanceSettings const &settings);
+
+  // TODO: Currently, SetCompositor API is not exposed to the developer due to threading issues.
```
Comment (acoates-ms) [inline vnext/Microsoft.ReactNative/ReactNativeAppBuilder.h:17]: What threading issues?

File: vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp (hunk 14 above)
Comment (sharath2727) [reply vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp:102]: Any suggestions for a default name? And yeah the title can updated anytime before the ReactNativeWin32::Start() is triggered, so the developers can change it later on.

File: vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp (hunk 15 above)
Comment (sharath2727) [reply vnext/Microsoft.ReactNative/ReactNativeAppBuilder.cpp:104]: Once the developer calls ReactNativeAppBuilder::Build() it returns a reactNativeWin32App object through which we can modify properties before calling ReactNativeWin32App::Start(). So to answer your question, yes apps can specify a custom size at any point of time. Thanks.

File: vnext/Microsoft.ReactNative/ReactNativeAppBuilder.h (hunk 16 above)
Comment (sharath2727) [reply vnext/Microsoft.ReactNative/ReactNativeAppBuilder.h]: Updated the comment to reflect.

File: packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp (hunk 13 above)
Comment (jonthysell) [reply packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp:33]: I still think I'd rather the you set the title/size on `reactNativeWin32App.AppWindow()` after build but before start to showcase *that* location as the first preference for where they customize.

File: packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp [hunk 17]
```diff
@@ -20,25 +20,7 @@ struct CompReactPackageProvider
...
-      window.ClientSize().Width / scaleFactor, window.ClientSize().Height / scaleFactor};
-  // Do not relayout when minimized
-  if (window.Presenter().as<winrt::Microsoft::UI::Windowing::OverlappedPresenter>().State() !=
-      winrt::Microsoft::UI::Windowing::OverlappedPresenterState::Minimized) {
-    winrt::Microsoft::ReactNative::LayoutConstraints constraints;
-    constraints.MaximumSize = constraints.MinimumSize = size;
-    rootView.Arrange(constraints, {0, 0});
-  }
-}
+// Scaling factor for the window's content based on the DPI of the display where the window is located.
```
Comment (jonthysell) [inline packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp:23]: is this comment even necessary?

File: packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp [hunk 18]
```diff
@@ -47,20 +29,11 @@ _Use_decl_annotations_ int CALLBACK WinMain(HINSTANCE instance, HINSTANCE, PSTR
...
-
   // Create a top-level window.
   auto window = winrt::Microsoft::UI::Windowing::AppWindow::Create();
   window.Title(windowTitle);
   window.Resize({1000, 1000});
   window.Show();
-  auto hwnd = winrt::Microsoft::UI::GetWindowFromWindowId(window.Id());
-  auto scaleFactor = ScaleFactor(hwnd);
 
   WCHAR appDirectory[MAX_PATH];
```
Comment (jonthysell) [inline packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp:38]: I wonder if it's possible to move this elsewhere too, maybe into another helper?

**Review by jonthysell — APPROVED** [review]

File: packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp (hunk 18 above)
Comment (sharath2727) [reply packages/sample-app-fabric/windows/SampleAppFabric/SampleAppFabric.cpp:38]: Sure I will take care of it in the following PRs.

**Review by jonthysell — APPROVED** [review]
